src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1573,21 +1576,61 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 if let Err(place_err) = self.is_mutable(place, is_local_mutation_allowed) {
                     error_reported = true;
 
-                    let item_msg = match self.describe_place(place) {
-                        Some(name) => format!("immutable item `{}`", name),
-                        None => "immutable item".to_owned(),
+                    let err_info = match *place_err {
+                        Place::Projection(ref proj) => {
```

Comment:
You can match on multiple levels to avoid deeply nested structures:

```rust
enum Foo {
    A { f: Bar },
    B(Bar),
}
enum Bar {
    X,
    Y,
}
let x = Foo::B(Bar::X);
match x {
    Foo::A { f: Bar::Y } => 1,
    _ => 0,
}
```